File: ak-framework/src/com/akframework/core/data/common/Entity.java
package com.akframework.core.data.common;

import com.akframework.core.data.annotation.Column;
import com.akframework.core.data.annotation.processor.DataAnnotationProcessor;
import com.akframework.kotlin.data.function.EntityFunctions;

import java.io.Serializable;
import java.lang.reflect.*;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static com.akframework.core.connection.ConnectionManager.*;

public abstract class Entity implements Serializable {

    @Column(value = "id")
    private int id;

    public int getId() {
        return id;
    }

    protected Optional<? extends Entity> manyToOne() {

        Optional<Entity> entity = Optional.empty();

        try {
            String methodName = Thread.currentThread().getStackTrace()[2].getMethodName();

            Method method = this.getClass().getDeclaredMethod(methodName);
            Map<String, String> manyToOneMap = EntityFunctions.scanManyToOne(method);

            DataAnnotationProcessor processor = EntityFunctions.createProcessor(this.getClass());
            String tableName = processor.getTable();

            String query = "SELECT * " +
                    "FROM " + manyToOneMap.get("referencedTable") + " " +
                    "WHERE id = (SELECT " + manyToOneMap.get("column") + " FROM " + tableName + " WHERE id = ?)";

            try (Connection connection = openConnection();
                 PreparedStatement statement = connection.prepareStatement(query)) {

                statement.setInt(1, this.getId());

                ResultSet resultSet = statement.executeQuery();

                if (resultSet.next()) {
                    Class<?> returnType = method.getReturnType();
                    Entity entityInstance = (Entity) returnType.getConstructor().newInstance();
                    List<Field> fieldList = EntityFunctions.loadFields(returnType);
                    EntityFunctions.setFields(entityInstance, fieldList, resultSet);
                    entity = Optional.of(entityInstance);
                }

                resultSet.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        }

        return entity;
    }

    protected List<? extends Entity> oneToMany() {

        List<Entity> entityList = new ArrayList<>();

        try {
            String methodName = Thread.currentThread().getStackTrace()[2].getMethodName();

            Method method = this.getClass().getDeclaredMethod(methodName);
            Map<String, String> oneToManyMap = EntityFunctions.scanOneToMany(method);

            DataAnnotationProcessor processor = EntityFunctions.createProcessor(this.getClass());
            String table = processor.getTable();

            String query = "SELECT * " +
                    "FROM " + oneToManyMap.get("table") + " " +
                    "WHERE " + oneToManyMap.get("referenceColumn") + " = (SELECT id FROM " + table + " WHERE id = ?)";

            try (Connection connection = openConnection();
                 PreparedStatement statement = connection.prepareStatement(query)) {

                statement.setInt(1, this.getId());

                ResultSet resultSet = statement.executeQuery();

                Type listType = method.getGenericReturnType();
                Type elementType = ((ParameterizedType) listType).getActualTypeArguments()[0];

                Class<?> elementClass = Class.forName(elementType.getTypeName());

                while (resultSet.next()) {
                    Entity entity = (Entity) elementClass.getConstructor().newInstance();
                    List<Field> fieldList = EntityFunctions.loadFields(elementClass);
                    EntityFunctions.setFields(entity, fieldList, resultSet);
                    entityList.add(entity);
                }
                resultSet.close();
            } catch (InstantiationException | InvocationTargetException |
                    SQLException | IllegalAccessException |
                    ClassNotFoundException e) {
                e.printStackTrace();
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        }

        return entityList;
    }
}

File: osiris-framework/src/main/com/osiris/kotlin/data/function/entity.kt
@file:JvmName("EntityFunctions")

package com.osiris.kotlin.data.function

import com.osiris.core.data.annotation.Column
import com.osiris.core.data.annotation.ManyToOne
import com.osiris.core.data.annotation.OneToMany
import com.osiris.core.data.annotation.processor.DataAnnotationProcessor
import com.osiris.core.data.common.Entity
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.sql.ResultSet
import java.sql.SQLException
import java.sql.Timestamp
import java.util.*

fun createProcessor(entityClass: Class<*>): DataAnnotationProcessor = DataAnnotationProcessor(entityClass)

fun loadFields(model: Class<*>): List<Field> {
    val fieldList = ArrayList<Field>()
    
    if (model.superclass != null) {
        fieldList.addAll(loadFields(model.superclass))
    }
    
    val modelFieldArray = model.declaredFields
    fieldList.addAll(Arrays.asList(*modelFieldArray))
    
    return fieldList
}

@Throws(SQLException::class, IllegalAccessException::class)
fun setFields(entity: Entity, fieldList: List<Field>, resultSet: ResultSet) {
    for (field in fieldList) {
        val column = field.getAnnotation(Column::class.java)
        if (column != null) {
            field.isAccessible = true
            val value: Any?
            val fieldType = field.type
            value = try {
                if (fieldType == Int::class.javaPrimitiveType || fieldType == Int::class.java) {
                    resultSet.getInt(column.value)
                } else if (fieldType == Boolean::class.javaPrimitiveType || fieldType == Boolean::class.java) {
                    resultSet.getBoolean(column.value)
                } else if (fieldType == Timestamp::class.java) {
                    resultSet.getTimestamp(column.value)
                } else if (fieldType == Double::class.javaPrimitiveType || fieldType == Double::class.java) {
                    resultSet.getDouble(column.value)
                } else {
                    resultSet.getString(column.value)
                }
            } catch (e: Exception) {
                null
            }
            
            field.set(entity, value)
        }
    }
}

fun scanManyToOne(method: Method): Map<String, String> {
    val map = HashMap<String, String>()
    
    val annotation = method.getAnnotation(ManyToOne::class.java)
    
    map.put("referencedTable", annotation.referencedTable)
    map.put("column", annotation.column)
    
    return map
}

fun scanOneToMany(method: Method): Map<String, String> {
    val map = HashMap<String, String>()
    
    val annotation = method.getAnnotation(OneToMany::class.java)
    
    map.put("table", annotation.table)
    map.put("referenceColumn", annotation.referenceColumn)
    
    return map
}
